Exposing translation status from partition translator

mmaslankaprv opened the pull request:
Exposed the `translation_status` from `partition_translator`. The `status` is used by the translation scheduler to determine the priority when the given partition translation is required. The status `next_checkpoint_deadline` is calculated as the time in which a current lag will exceed the target lag if no translation for the partition happens.  

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
dl/manager: added ntp to log about failure in initializing datasource

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
dl/translation: expose buffered_bytes from translation_context

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62858-->
<!--1c2596ba-bbca-471e-98bd-e89d68f85849-->
<!--612d9067-c807-42cb-a087-231ee730b3e2-->
<!--f9b053d8-8e89-4b25-af79-fc8bba8d04cb-->
<!--0d0d4735-3553-4464-8a17-f5786663f89d-->
<!--2811924d-8706-4b8d-b831-ac2df832b726-->
<!--6449a966-d018-4fb4-aa40-9bf14d3d6c45-->
<!--title-->
## Retry command for [Build#62858](https://buildkite.com/redpanda/redpanda/builds/62858)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"nessie","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a8bc0bc3-f954-4d45-af28-391154b5ff97-->
<!--TEST RESULTS-->
<!--82edd09c-3cf1-4d15-90fc-869f0829c60d-->
<!--TEST RESULTS-->
<!--212bb1cc-de7d-4b3b-9783-ff724b6d8d5f-->
<!--TEST RESULTS-->
<!--2049af91-3db5-45c3-8a89-f99ffe216073-->
<!--TEST RESULTS-->
<!--145f43f4-9f32-40fc-9eed-b04fb3c0389a-->
<!--TEST RESULTS-->
<!--7e6c0477-8b22-4522-be89-9c85ca78a6b8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62858">build#62858</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580ef-1333-4d69-bf56-25417fd3b052 | FLAKY | 1/14 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580f3-1da3-45c0-8599-8bb0c5d0320c | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580ef-1334-4147-a659-9d909e2e5cdb | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580f3-1da4-45cd-86d9-2ac0fa43b849 | FLAKY | 1/7 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580ef-1334-47b6-9a50-23e39f2f1a7a | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580f3-1da1-4055-a0f7-0b681e4979e6 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580ef-1335-410d-b081-5286eee5694e | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580f3-1da2-41e6-a761-a6b3665397c5 | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580ef-1333-4d69-bf56-25417fd3b052 | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580f3-1da3-45c0-8599-8bb0c5d0320c | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580ef-1334-4147-a659-9d909e2e5cdb | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580f3-1da4-45cd-86d9-2ac0fa43b849 | FAIL | 0/20 |
| rptest.tests.full_disk_test.FullDiskReclaimTest.test_full_disk_triggers_gc | ducktape | https://buildkite.com/redpanda/redpanda/builds/62858#019580ef-1334-4147-a659-9d909e2e5cdb | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62930">build#62930</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62930#0195863e-30c8-4f16-8243-acb0ea85dfc9 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#0195869d-388e-4014-b395-93a3b9f25d63 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#01958699-addc-4de3-84dd-276f2c34748f | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#0195869d-388f-456e-a83a-d9f72fda0d86 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#0195869d-388c-448c-91e3-e30950b63f2d | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#01958699-adde-49dc-9e48-2050f860200a | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#0195869d-388d-41f4-b6e4-09f3fa6895e8 | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#01958699-addb-4378-bb13-ed484a447689 | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#0195869d-388e-4014-b395-93a3b9f25d63 | FLAKY | 1/11 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#01958699-addc-4de3-84dd-276f2c34748f | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#0195869d-388f-456e-a83a-d9f72fda0d86 | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#01958699-addc-4de3-84dd-276f2c34748f | FLAKY | 1/2 |
| rptest.tests.mirror_maker_test.TestMirrorMakerService.test_simple_end_to_end.source_type=redpanda | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#01958699-addc-4de3-84dd-276f2c34748f | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62930#0195869d-388d-41f4-b6e4-09f3fa6895e8 | FLAKY | 1/2 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62930#0195863e-30c8-4f16-8243-acb0ea85dfc9 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62975">build#62975</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=False.server_cfg=300000.topic_cfg=None | ducktape | https://buildkite.com/redpanda/redpanda/builds/62975#019589c6-3955-49c4-9d5b-8d84c53f535a | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63068">build#63068</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/63068#019590ea-464d-43c5-8338-8a3e4dd0a22c | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63068#019590e6-a56e-4ee7-b849-38b0baf73795 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63068#019590ea-4649-4c36-bedf-41fa8242163f | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63115">build#63115</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63115#019593c0-a0e1-4459-addd-350f423f6b30 | FLAKY | 1/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/63115#019593c0-a0e2-4b9b-b270-2efa119df535 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63115#019593c4-3844-4eac-b18f-35bf154107a7 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63142">build#63142</a></summary>


| test_id                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63142#01959532-5057-4783-9905-8f3fad34c01d | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63142#01959534-f940-4bd4-bfe6-b9aeb24e1535 | FLAKY         | 1/2      |
</details>
<!--/body-->

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62930-->
<!--e4b1e98c-9fff-42f2-a05c-7ea4c549eae1-->
<!--855939e6-1410-46a4-b273-f7e22dc9dbde-->
<!--a559c050-6f72-4db1-84d8-e642d479be64-->
<!--c93eef70-c7e2-499f-8e93-00079c7d2def-->
<!--0552c273-9893-4440-a755-bd92ec11f29e-->
<!--title-->
## Retry command for [Build#62930](https://buildkite.com/redpanda/redpanda/builds/62930)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"nessie","cloud_storage_type":1}
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_restore_partition_spec@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
```
<!--/body-->

mmaslankaprv commented:
/dt

mmaslankaprv commented:
I am aware of the last `wip` commit, will remove it as soon as we merge Andrew changes

oleiman commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -367,7 +367,10 @@ datalake_manager::handle_translator_state_change(const model::ntp& ntp) {
-        vlog(datalake_log.warn, "adding translator failed, retrying in a bit");
+        vlog(
+          datalake_log.warn,
+          "adding translator for {} failed, retrying in a bit",
```
nitpick: while we're in here, should we either specify the delay or remove "in a bit"? the ambiguous wording in a WARN level log feels a bit strange.

oleiman commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -335,6 +335,23 @@ datalake_manager::handle_translator_state_change(const model::ntp& ntp) {
     auto data_src = translation::data_source::make_default_data_source(
       partition);
+
+    // TODO: make the initialization timeout configurable
```
> configurable

just curious - like cluster or topic property? and is it really necessary / what is the benefit?

oleiman commented on src/v/datalake/translation/deps.cc:
```diff
@@ -303,19 +314,56 @@ class partition_data_source : public data_source {
+
+        auto size_after_max_translatable
+          = _partition->log()->size_bytes_after_offset(
+            _partition->last_stable_offset());
```
is there a reason not to use `max_offset_for_translation` here? I see the semantics are slightly different, but I don't understand why we would want one or the other, exactly.

andrwng replied:
My guess is that it's a kafka::offset instead of a RP offset

mmaslankaprv replied:
exactly, for the size calculation we need `model::offset`

oleiman replied:
oh, duh. thanks

oleiman commented on src/v/datalake/translation/deps.cc:
```diff
@@ -343,6 +391,12 @@ class partition_data_source : public data_source {
     // an offset meeting or exceeding target.offset.
     std::optional<timestamped_offset> _translation_target;
+    /**
+     * Cache values in order to avoid repeated calls to the STM, the cached
+     * value is updated on initialization and every time the STM is updated with
+     * the new translated offset.
```
looks like we only update this at the start of the loop (i.e. after reconciling w/ the coordinator), but not after the STM update following a successful translation checkpoint. is that intentional?

bharathv replied:
> Cache value in order to avoid repeated calls to the STM

q: it seems stm->highest_translated_offset() is inexpensive since the stm also caches it, do we need this additional caching?

mmaslankaprv replied:
I added caching as the `state_machine::highest_translated_offset` requires a `sync` and it returns a future. All the status related interfaces are synchronous

oleiman commented on src/v/datalake/translation/tests/partition_translator_tests.cc:
```diff
@@ -95,6 +95,10 @@ class fake_test_ctx {
           _highest_translated_offset + kafka::offset_delta(100));
     }
 
+    std::optional<size_t> translation_backlog() const {
+        return _translation_backlog;
+    }
```
worth adding a unit test of some kind?

oleiman reviewed (commented):
handful of questions, but looks sensible to me.

andrwng commented on src/v/datalake/translation/deps.h:
```diff
@@ -94,6 +94,10 @@ class data_source {
+    /**
+     * Initializes data source with offsets from the STM.
+     */
+    virtual ss::future<> initialize(model::timeout_clock::duration) = 0;
```
nit: maybe add `timeout` as a name so it's clear what it's used for

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -164,6 +165,16 @@ class partition_data_source : public data_source {
           std::bind_front(&wait_stm_translated, _stm))) {}
 
+    ss::future<> initialize(model::timeout_clock::duration timeout) final {
+        auto highest_translated_offset
+          = co_await _stm->highest_translated_offset(timeout);
```
Curious, if this times out should we retry or something? I guess it will just mean that that backlog computation won't be correct

mmaslankaprv replied:
The timeout is only possible if `stm` sync fails, if it does that will effectively prevent translation from happening

mmaslankaprv replied:
Actually i will add retries

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -325,7 +328,7 @@ ss::future<> partition_translator::init(
         });
     });
-    return ss::make_ready_future();
+    _cached_current_lag = co_await current_lag_ms();
```
Curious if there's harm in making this the first thing we do? Then we at least start translating with a cached value in hand? Or is the idea that using an uninitialized cached lag is fine

andrwng commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -166,6 +166,7 @@ class partition_translator : public scheduling::translator {
 
     // duration and abort source for a scheduled translation
     std::optional<inflight_translation_state> _inflight_translation_state;
+    std::optional<scheduling::clock::duration> _cached_current_lag;
```
Would be great if you could add a comment explaining how callers should treat this value when it's null.

Also curious if it would make sense to look directly at the stm without the sync() calls or somesuch instead of caching?

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -201,7 +201,10 @@ ss::future<> partition_translator::translate_until_stopped() {
+        /**
+         * Update cached lag
+         */
+        _cached_current_lag = co_await current_lag_ms();
```
What's the rationale behind caching the lag rather than caching the last translated timestamp? I would've thought that we care more about the timestamp since the actual value for lag will change between here and the next call to `status()`

mmaslankaprv committed:
dl/translation_stm: added methods returning cached state

It not always critical to get the most up do date synced state from the
`datalake::translation::state_machine`. Added synchronous methods
accessing the latest translated offset and timestamp without syncing the
stm first.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
dl/translation: reorganized lack tracking mechanism

Reorganized the lag tracking mechanism so that all lag information are
stored in the `lag_tracker`.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
dl/manager: added log entry when translator state changes

Signed-off-by: Michał Maślanka <michal@redpanda.com>

oleiman commented on src/v/datalake/translation/deps.cc:
```diff
@@ -607,40 +585,156 @@ class partition_lag_tracker : public translation_lag_tracker {
+
+    scheduling::clock::time_point next_checkpoint_deadline() const final {
+        auto time_left = std::chrono::milliseconds(
+                           static_cast<int>(0.9 * target_lag().count()))
```
nitpick: this `0.9 * target_lag` snippet appears in two places. would be nice to at least bind the 0.9 portion to an identifier.

bharathv replied:
cast to milliseconds::rep instead of int?

oleiman commented on src/v/datalake/translation/deps.h:
```diff
@@ -109,6 +109,11 @@ class data_source {
      * min_offset_for_translation, next(last_translated_offset) if one exists,
      * or nullopt if there is no data to translate.
+     * TODO:
+     * if we decide not to return on any data appearing we need to add a
+     * callback to the method ss::noncopyable_function<void(kafka::offset)>
+     * that would notify the lag tracker about new offsets that are ready to be
+     * translated.
```
q: is this in reference to the "wait for enough data to appear" todo?

mmaslankaprv replied:
exactly

oleiman commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -202,12 +203,15 @@ ss::future<> partition_translator::translate_until_stopped() {
+        auto lto_from_coordinator = result.last_added_offset.value_or(
           kafka::prev_offset(_data_source->min_offset_for_translation()));
-
+        // TODO: maybe we should store the translation timestamp in
+        // coordinator
+        // ?
```
I had added some of this new state to the coordinator before, and it felt out of place since these timestamps are not an absolute quantity with respect to either the catalog or the source ntp. 

if a translator comes up and there's no "last timestamp" there, that doesn't effect the semantics of translation; we will just translate some and generate a timestamp ad hoc.

I don't feel too strongly about it though, just my $0.02.

oleiman commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -320,17 +318,30 @@ ss::future<> partition_translator::translate_until_stopped() {
               _logger.warn,
               "error updating highest translated offset: {}",
-              reset_error);
+              result);
```
```suggestion
              replicate_result);
```

oleiman commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -391,16 +402,17 @@ ss::future<> partition_translator::close() noexcept {
 ss::future<std::optional<std::chrono::milliseconds>>
 partition_translator::current_lag_ms() const {
-    return _data_source->current_lag_ms(wait_timeout);
+    co_return _lag_tracking->current_lag_ms();
```
nitpick: any reason this still needs to return a future?

mmaslankaprv replied:
no, will change this

oleiman reviewed (commented):
lgtm. all nits & q's with the exception of a mistyped log format parameter.

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -153,6 +154,63 @@ coordinator_api::make_default_coordinator_api(coordinator::frontend& frontend) {
+}
+
+struct timestamped_offset {
+    timestamped_offset() = default;
```
Curious if this is actually used?

bharathv replied:
L737 -- ah you mean the default c-tor 😶‍🌫️

andrwng replied:
Isn't that just initialized as a nullopt?

bharathv replied:
ya it its, I thought you were asking if `timestamped_offset` is ever used, my bad.

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -320,17 +318,30 @@ ss::future<> partition_translator::translate_until_stopped() {
+        const auto translated_offset_ts
+          = _lag_tracking->get_translated_offset_timestamp(
+            last_translated_offset);
+        _logger.trace(
+          "Replicating translation checkpoint with offset: {} and timestamp: "
+          "{}",
+          last_translated_offset,
+          translated_offset_ts);
```
I thought we would only want to replicate the translated timestamp if we actually caught up to the target offset?

oleiman replied:
oh yeah good catch.

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -607,40 +585,156 @@ class partition_lag_tracker : public translation_lag_tracker {
+    }
+
+    std::optional<model::timestamp>
+    get_translated_offset_timestamp(kafka::offset) final {
```
Shouldn't this check if the input offset is >= the translation target offset?

bharathv replied:
Also there is a risk of dereferencing null optional in L716

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63068-->
<!--16bef60e-f378-45e4-b465-777493de0bff-->
<!--title-->
## Retry command for [Build#63068](https://buildkite.com/redpanda/redpanda/builds/63068)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_upgrade_test.py::DatalakeUpgradeTest.test_upload_through_upgrade@{"cloud_storage_type":1,"query_engine":"spark"}
```
<!--/body-->

bharathv commented on src/v/datalake/translation/state_machine.h:
```diff
@@ -65,6 +65,24 @@ class translation_stm final : public raft::persisted_stm<> {
 
+    /**
+     * Cached versions of highest_translated_offset method. These methods will
+     * not call state machine sync, risking to have she stale value if the
```
nit: typo "she"

(same below)

bharathv commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -290,6 +290,7 @@ datalake_manager::handle_translator_state_change(const model::ntp& ntp) {
     if (_gate.is_closed() || !model::is_user_topic(ntp)) {
         co_return;
     }
+    vlog(datalake_log.debug, "Translator state change for {} detected.", ntp);
```
thank you.

bharathv commented on src/v/datalake/translation/deps.cc:
```diff
@@ -153,6 +154,63 @@ coordinator_api::make_default_coordinator_api(coordinator::frontend& frontend) {
+}
+std::chrono::milliseconds calculate_target_lag(
+  const cluster::topic_table& topics, const model::ntp& ntp) {
+    // todo: In addition to integrating with config subsystem, an additional
+    // task that needs to be done is to keep track of translation intervals
+    // centrally some place (coordinator?). This is needed so that we do not
+    // lose the interval boundaries due to a crash. For example imagine
+    // keeping it in memory and target lag is set to 1d and the leadership
+    // transfer happens after half a day, we cannot just reset the interval
+    // window on the new leader. That probably works for smaller lag values
```
Think this was moved from earlier version of the code.. AFAICT this is not needed anymore?

bharathv commented on src/v/datalake/translation/deps.cc:
```diff
@@ -601,40 +585,156 @@ class partition_lag_tracker : public translation_lag_tracker {
       ss::lw_shared_ptr<cluster::partition> partition,
       cluster::topic_table& topics)
       : _partition(std::move(partition))
+      , _partition_proxy(std::make_unique<kafka::partition_proxy>(
```
nit: todo in L577 can be removed too.

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -202,12 +203,15 @@ ss::future<> partition_translator::translate_until_stopped() {
+        auto lto_from_coordinator = result.last_added_offset.value_or(
           kafka::prev_offset(_data_source->min_offset_for_translation()));
-
+        // TODO: maybe we should store the translation timestamp in
```
nit: may be add a little more detail for future

mmaslankaprv committed:
dl/translation: make get_target_lag method not returning a future

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63115-->
<!--098697bb-20d2-4dc5-80c9-ee28889adfa6-->
<!--title-->
## Retry command for [Build#63115](https://buildkite.com/redpanda/redpanda/builds/63115)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_upgrade_test.py::DatalakeUpgradeTest.test_upload_through_upgrade@{"cloud_storage_type":1,"query_engine":"spark"}
```
<!--/body-->

mmaslankaprv commented:
Latest force push: fixed offset gap detection

oleiman commented:
> Latest force push: fixed offset gap detection

Having a bit of trouble seeing what changed, logically. What was the problem with offset gap detection? (I assume this means a mismatch between the coordinator-replicated LTO vs translator local LTO)
